apps/keyboard/js/imes/latin/predictions.js

```diff
@@ -418,29 +418,53 @@ var Predictions = function() {
     }
 
     // Check whether all the characters of s appear in the dictionary or are
-    // at least near characters that do. If we are passed a string that does
-    // not pass this test then there is no way we will be able to offer
-    // suggestions and it is not even worth searching.
+    // at least their variant forms or nearby characters that do. If we are
+    // passed a string that does not pass this test then there is no way we will
+    // be able to offer suggestions and it is not even worth searching.
     function validChars(s) {
-      outer: for (var i = 0, n = s.length; i < n; i++) {
-        var c = s.charCodeAt(i);
-        if (characterTable.hasOwnProperty(c))  // character is valid
-          continue;
-        // If the character does not occur in this language, but there is
-        // a nearby key that does occur, then maybe it is okay
-        if (!nearbyKeys.hasOwnProperty(c))
-          return false;  // no nearby keys, so no suggestions possible
-        var nearby = nearbyKeys[c];
-        for (c in nearby) {
-          if (characterTable.hasOwnProperty(c))
-            continue outer;
+      return Array.from(s).every(function(c) {
```

Comment:
This seems terribly inefficient. Looping over every character and then for every char we loop over the variant table and then do another indexOf. Can you make a new lookup table (during initialization of the worker or something) where you flatten this data? Then this should be an O(1) operation.